### ISSUE TITLE
IAM role masterAnnotation results in continuous jenkins pod restart

### ISSUE THREAD

**Issue by BogdanSorlea** [jenkinsci/kubernetes-operator#229]
Hey,
So adding just these lines (what's between `[[ ]]` is given just for context):
```
[[
apiVersion: jenkins.io/v1alpha2
kind: Jenkins
spec:
  master:
]]
    masterAnnotations:
      iam.amazonaws.com/role: <IAM_ROLE_NAME>
```
results in the Jenkins pod being killed over and over by the operator.

The operator's logs look like this:
```
2019-12-11T13:09:36.277Z	INFO	controller-jenkins	base/reconcile.go:544	Jenkins pod annotations have changed to 'map[iam.amazonaws.com/role:<IAM_ROLE_NAME>]', recreating pod	{"cr": "can-ci"}
2019-12-11T13:09:36.277Z	INFO	controller-jenkins	base/reconcile.go:693	Terminating Jenkins Master Pod default/jenkins-can-ci	{"cr": "can-ci"}
2019-12-11T13:09:48.628Z	INFO	controller-jenkins	base/reconcile.go:437	Creating a new Jenkins Master Pod default/jenkins-can-ci	{"cr": "can-ci"}
2019-12-11T13:09:49.931Z	INFO	controller-jenkins	base/reconcile.go:719	Jenkins master pod starting timeout, events '[Message: Readiness probe failed: Get http://10.100.3.96:8080/login: dial tcp 10.100.3.96:8080: connect: connection refused Subobject: spec.containers{jenkins-master}]'	{"cr": "can-ci"}
2019-12-11T13:09:49.931Z	INFO	controller-jenkins	jenkins/jenkins_controller.go:352	Setting operator version	{"cr": "can-ci"}
2019-12-11T13:11:18.688Z	INFO	controller-jenkins	base/reconcile.go:544	Jenkins pod annotations have changed to 'map[iam.amazonaws.com/role:<IAM_ROLE_NAME>]', recreating pod	{"cr": "can-ci"}
2019-12-11T13:11:18.688Z	INFO	controller-jenkins	base/reconcile.go:693	Terminating Jenkins Master Pod default/jenkins-can-ci	{"cr": "can-ci"}
2019-12-11T13:11:19.047Z	INFO	controller-jenkins	base/reconcile.go:544	Jenkins pod annotations have changed to 'map[iam.amazonaws.com/role:<IAM_ROLE_NAME>]', recreating pod	{"cr": "can-ci"}
2019-12-11T13:11:19.047Z	INFO	controller-jenkins	base/reconcile.go:693	Terminating Jenkins Master Pod default/jenkins-can-ci	{"cr": "can-ci"}
2019-12-11T13:11:20.257Z	INFO	controller-jenkins	base/reconcile.go:544	Jenkins pod annotations have changed to 'map[iam.amazonaws.com/role:<IAM_ROLE_NAME>]', recreating pod	{"cr": "can-ci"}
2019-12-11T13:11:20.257Z	INFO	controller-jenkins	base/reconcile.go:693	Terminating Jenkins Master Pod default/jenkins-can-ci	{"cr": "can-ci"}
2019-12-11T13:11:21.460Z	INFO	controller-jenkins	base/reconcile.go:544	Jenkins pod annotations have changed to 'map[iam.amazonaws.com/role:<IAM_ROLE_NAME>]', recreating pod	{"cr": "can-ci"}
2019-12-11T13:11:21.460Z	INFO	controller-jenkins	base/reconcile.go:693	Terminating Jenkins Master Pod default/jenkins-can-ci	{"cr": "can-ci"}
2019-12-11T13:11:22.712Z	INFO	controller-jenkins	base/reconcile.go:544	Jenkins pod annotations have changed to 'map[iam.amazonaws.com/role:<IAM_ROLE_NAME>]', recreating pod	{"cr": "can-ci"}
2019-12-11T13:11:22.712Z	INFO	controller-jenkins	base/reconcile.go:693	Terminating Jenkins Master Pod default/jenkins-can-ci	{"cr": "can-ci"}
2019-12-11T13:11:23.842Z	INFO	controller-jenkins	base/reconcile.go:544	Jenkins pod annotations have changed to 'map[iam.amazonaws.com/role:<IAM_ROLE_NAME>]', recreating pod	{"cr": "can-ci"}
```

the logs of the jenkins pod around the time of death (probably not relevant anyways):
```
Downloading plugin: apache-httpcomponents-client-4-api from https://updates.jenkins.io/2.176/latest/apache-httpcomponents-client-4-api.hpi
 > plain-credentials depends on credentials:2.1.16
 > workflow-step-api depends on structs:1.20
 > job-dsl depends on structs:1.19,script-security:1.54,vsphere-cloud:1.1.11;resolution:=optional,config-file-provider:2.15.4;resolution:=optional,managed-scripts:1.3;resolution:=optional,configuration-as-code:1.15;resolution:=optional
Skipping optional dependency vsphere-cloud
Downloading plugin: script-security from https://updates.jenkins.io/2.176/latest/script-security.hpi
Skipping optional dependency config-file-provider
Skipping optional dependency managed-scripts
Skipping optional dependency configuration-as-code
 > workflow-api depends on workflow-step-api:2.16,scm-api:2.2.6,structs:1.17
Downloading plugin: scm-api from https://updates.jenkins.io/2.176/latest/scm-api.hpi
rpc error: code = Unknown desc = Error: No such container: 5e40f4449864087d12950c984c9a4b6f82ba94ea9d5fbb0511b34dd45364c51c
```

So I'm not sure what is going on here and what I might be missing, but it's certain that this change is the culprit and that it works well before adding that annotation.

Also, is this annotation supposed to take in the role name or the instance-profile name? I seem to have learnt that it needs the role, but I might be wrong(?) In any case, my instance-profile has a different name than the role. I might try that out.

Other refs: I've briefly mentioned this problem in another ticket, but there's no relevant debug info there (will link here once I find it). Also, it might be a problem similar to what was going on here: https://github.com/jenkinsci/kubernetes-operator/issues/38

**Comment by tomaszsek** [comment]
Hi @BogdanSorlea,

Could you provide the output from command `kubectl -n jenkins get po jenkins-example -o yaml`(let suppose the namespace is `jenkins` and Jenkins CR is `example`)?

Cheers

**Comment by BogdanSorlea** [comment]
hey @tomaszsek 
here you go:
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/podIP: 10.100.3.205/32
    iam.amazonaws.com/role: <IAM_ROLE_NAME>
  creationTimestamp: 2019-12-11T14:17:58Z
  deletionGracePeriodSeconds: 30
  deletionTimestamp: 2019-12-11T14:18:32Z
  labels:
    app: jenkins-operator
    jenkins-cr: can-ci
  name: jenkins-can-ci
  namespace: default
  ownerReferences:
  - apiVersion: jenkins.io/v1alpha2
    blockOwnerDeletion: true
    controller: true
    kind: Jenkins
    name: can-ci
    uid: f844adb9-1a8c-11ea-87f2-0a8c7d2b1ea0
  resourceVersion: "96115472"
  selfLink: /api/v1/namespaces/default/pods/jenkins-can-ci
  uid: 082b06b9-1c21-11ea-87f2-0a8c7d2b1ea0
spec:
  containers:
  - command:
    - bash
    - -c
    - /var/jenkins/scripts/init.sh && /sbin/tini -s -- /usr/local/bin/jenkins.sh
    env:
    - name: JENKINS_HOME
      value: /var/jenkins/home
    - name: JAVA_OPTS
      value: -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -XX:MaxRAMFraction=1
        -Djenkins.install.runSetupWizard=false -Djava.awt.headless=true -Djavax.net.ssl.trustStore=/var/jenkins/home/.cacerts/trust-store.jks
        -Djavax.net.ssl.trustStorePassword=somepasswordthisisjustadummyanditsnotitdummy
    - name: JENKINS_OPTS
      value: --sessionTimeout=240 --sessionEviction=7200
    image: docker.ourserver.net/jenkins:lts
    imagePullPolicy: Always
    livenessProbe:
      failureThreshold: 12
      httpGet:
        path: /login
        port: 8080
        scheme: HTTP
      initialDelaySeconds: 120
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 5
    name: jenkins-master
    ports:
    - containerPort: 8080
      name: http
      protocol: TCP
    - containerPort: 50000
      name: slavelistener
      protocol: TCP
    readinessProbe:
      failureThreshold: 12
      httpGet:
        path: /login
        port: 8080
        scheme: HTTP
      initialDelaySeconds: 90
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 5
    resources:
      limits:
        cpu: "1"
        memory: 2Gi
      requests:
        cpu: "1"
        memory: 500Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/jenkins/home
      name: jenkins-home
    - mountPath: /var/jenkins/scripts
      name: scripts
      readOnly: true
    - mountPath: /var/jenkins/init-configuration
      name: init-configuration
      readOnly: true
    - mountPath: /var/jenkins/operator-credentials
      name: operator-credentials
      readOnly: true
    - mountPath: /var/jenkins/home/.cacerts
      name: trust-store
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: jenkins-operator-can-ci-token-fhzk9
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: ip-10-80-58-231.eu-west-1.compute.internal
  priority: 0
  restartPolicy: Never
  schedulerName: default-scheduler
  securityContext:
    fsGroup: 1000
    runAsUser: 1000
  serviceAccount: jenkins-operator-can-ci
  serviceAccountName: jenkins-operator-can-ci
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir: {}
    name: jenkins-home
  - configMap:
      defaultMode: 511
      name: jenkins-operator-scripts-can-ci
    name: scripts
  - configMap:
      defaultMode: 420
      name: jenkins-operator-init-configuration-can-ci
    name: init-configuration
  - name: operator-credentials
    secret:
      defaultMode: 420
      secretName: jenkins-operator-credentials-can-ci
  - configMap:
      defaultMode: 420
      name: trust-store
    name: trust-store
  - name: jenkins-operator-can-ci-token-fhzk9
    secret:
      defaultMode: 420
      secretName: jenkins-operator-can-ci-token-fhzk9
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2019-12-11T14:18:00Z
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2019-12-11T14:18:00Z
    message: 'containers with unready status: [jenkins-master]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2019-12-11T14:18:00Z
    message: 'containers with unready status: [jenkins-master]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: 2019-12-11T14:17:58Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://8b6f06356d3c82450138fc49a61bf5721429f276962cdb6c8596e0ce779eab53
    image: docker.ourserver.net/jenkins:lts
    imageID: docker-pullable://docker.ourserver.net/jenkins@sha256:825b3802929a34a8c135e7fb19d72c30845f469bc7f9d42eeaf2f290167c9f1f
    lastState: {}
    name: jenkins-master
    ready: false
    restartCount: 0
    state:
      running:
        startedAt: 2019-12-11T14:18:02Z
  hostIP: 10.80.58.231
  phase: Running
  podIP: 10.100.3.205
  qosClass: Burstable
  startTime: 2019-12-11T14:18:00Z
```

**Comment by pawelprazak** [comment]
That's weird, I use `masterAnnotations` with `iam.amazonaws.com/role` in it and it works.
My environment is on AWS EKS btw

**Comment by tomaszsek** [comment]
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/podIP: 10.100.3.205/32
    iam.amazonaws.com/role: <IAM_ROLE_NAME>
```
You have to add `cni.projectcalico.org/podIP: 10.100.3.205/32` to the Jenkins Cr as well.

**Comment by pawelprazak** [comment]
It's probably added by Calico automatically and will change with every restart.

**Comment by pawelprazak** [comment]
btw @tomaszsek we need to deprecate the `masterAnnotations` in favour of a better name or maybe the whole metadata sections or something.

**Comment by pawelprazak** [comment]
also it might be related to #197

**Comment by BogdanSorlea** [comment]
> You have to add `cni.projectcalico.org/podIP: 10.100.3.205/32` to the Jenkins CR as well.

can you be more specific / provide an example?

**Comment by BogdanSorlea** [comment]
a somewhat unrelated comment:

in relation to this comment - https://github.com/jenkinsci/kubernetes-operator/issues/185#issuecomment-549267007 - while I understand that the context might be related to a (temporary) workaround, I feel I should stress that (possibly) no solution (for any problem or for design questions) should make assumptions about the namespace in which jenkins is running - but I'm sure you probably have that in agreement already

**Comment by pawelprazak** [comment]
So what Tomek was suggesting is something like:
```
  master:
    masterAnnotations:
      iam.amazonaws.com/role: <IAM_ROLE_NAME>
      cni.projectcalico.org/podIP: 10.100.3.205/32
```
but it's more to confirm that it's what we think it is then as a viable workaround.

To clarify few things:
- there are no assumptions as to what namespace jenkins operator runs in
- calico is your CNI and is not related to `kube2iam`

The #197 tracks what we think will solve this issue as well as issues we have with every admission controller, like in Istio or OpenShift default security settings. Now we can add Calico to the list.

**Comment by BogdanSorlea** [comment]
@pawelprazak @tomaszsek 
I added the `cni.projectcalico.org/podIP: 10.100.3.205/32` annotation as well, but the behaviour is the same, the pod keeps getting restarted - and upon inspecting the yaml describing the pod, that calico IP keeps changing with each restart.

On a related note, deploying without any masterAnnotations and inspecting the yaml describing the pod would reveal the `cni.projectcalico.org/podIP` annotation to be present, so I therefore understand that it's more the case of whether masterAnnotations are present on Jenkins at all (in which case the jenkins-operator would work the reconcile, otherwise, if missing, it would ignore pod annotations altogether) rather than a problem with the kube2iam annotation. Which I guess is what you pointed out when mentioning that calico is the CNI / is not related to `kube2iam`.

So anyways, I at least understand the problem better now.

Any forecast on when, in time, the problem might be fixed (in the sense of #197)? Any chance for a workaround in the meantime - like annotations that the reconcile script should ignore - given that just adding the annotation does not make it work?

**Comment by pawelprazak** [comment]
Thanks, that confirms our hypothesis, this is what I predicted will happen, given the IP is dynamic.

I'm afraid no workaround is available at the moment, as to when #197 might be done, I think we'll raise it's priority.

**Comment by tomaszsek** [comment]
Fixed in https://github.com/jenkinsci/kubernetes-operator/releases/tag/v0.3.0
